postgres/datadog_checks/postgres/connections.py

```diff
@@ -86,51 +88,50 @@ def _get_connection_raw(
         dbname: str,
         ttl_ms: int,
         timeout: int = None,
-        startup_fn: Callable[[psycopg2.extensions.connection], None] = None,
+        startup_fn: Callable[[psycopg.Connection], None] = None,
         persistent: bool = False,
-    ) -> psycopg2.extensions.connection:
+    ) -> psycopg.Connection:
         """
-        Return a connection from the pool.
+        Return a connection pool for the requested database from the managed pool.
         Pass a function to startup_func if there is an action needed with the connection
         when re-establishing it.
         """
+        if not timeout:
+            # We don't want the wait indefinitely for the connection pool to free up
+            # if timeout is not specified, so we set it to the default connection timeout
+            timeout = self._config.connection_timeout
         start = datetime.datetime.now()
         self.prune_connections()
-        with self._mu:
-            conn = self._conns.pop(dbname, ConnectionInfo(None, None, None, None, None, None))
-            db = conn.connection
-            if db is None or db.closed:
-                if self.max_conns is not None:
-                    # try to free space until we succeed
-                    while len(self._conns) >= self.max_conns:
-                        self.prune_connections()
-                        self.evict_lru()
-                        if timeout is not None and (datetime.datetime.now() - start).total_seconds() > timeout:
-                            raise ConnectionPoolFullError(self.max_conns, timeout)
-                        time.sleep(0.01)
-                        continue
-                self._stats.connection_opened += 1
-                db = self.connect_fn(dbname)
-                if startup_fn:
-                    startup_fn(db)
-            else:
-                # if already in pool, retain persistence status
-                persistent = conn.persistent
-
-            if db.status != psycopg2.extensions.STATUS_READY:
-                # Some transaction went wrong and the connection is in an unhealthy state. Let's fix that
-                db.rollback()
-
-            deadline = datetime.datetime.now() + datetime.timedelta(milliseconds=ttl_ms)
-            self._conns[dbname] = ConnectionInfo(
-                connection=db,
-                deadline=deadline,
-                active=True,
-                last_accessed=datetime.datetime.now(),
-                thread=threading.current_thread(),
-                persistent=persistent,
-            )
-            return db
+        conn = self._conns.pop(dbname, ConnectionInfo(None, None, None, None, None))
+        db = conn.connection
+        if db is None or db.closed or db.info.status != psycopg.pq.ConnStatus.OK:
+            # db.info.status checks if the connection is still active on the server
+            if self.max_conns is not None:
```

Comment:
I think this max conns logic needs to be re-worked a bit now that we are instantiating one of these per thread... Does it make sense to get rid of this part of the feature for now?

I think the issue would mainly manifest in the explain plan code path and the schema collection/ autodiscovery code paths, which would hold open a connection to every database it:

sees queries running so it can explain them
discovered through autodiscovery and we want to grab relation metrics/schema stuff from